forward/rewriting.lua

```diff
@@ -204,31 +204,42 @@ function Rewriting:process_packet(i, o)
       P.free(p)
       return
    end
-   if backend_len == 16 then
-      error("ipv6 output not implemented")
-   elseif backend_len ~= 4 then
-      error("backend length must be 4 (for ipv4) or 16 (for ipv6)")
-   end
+
 
    -- unparse L4 and L3
    datagram:unparse(2)
 
    local gre_header = GRE:new({protocol = l3_type})
    datagram:push(gre_header)
 
-   local outer_ip_header = IPV4:new({src = self.ipv4_addr,
-                                     dst = backend,
-                                     protocol = L4_GRE,
-                                     ttl = self.ttl})
-   outer_ip_header:total_length(
-      ip_total_length + gre_header:sizeof() + outer_ip_header:sizeof())
-   -- need to recompute checksum after changing total_length
-   outer_ip_header:checksum()
+
+   local outer_ip_header, outer_l3_type
+   if backend_len == 4 then -- IPv4
+      outer_l3_type = L3_IPV4
+      outer_ip_header = IPV4:new({src = self.ipv4_addr,
+                                        dst = backend,
+                                        protocol = L4_GRE,
+                                        ttl = self.ttl})
+      outer_ip_header:total_length(
+         ip_total_length + gre_header:sizeof() + outer_ip_header:sizeof())
+      -- need to recompute checksum after changing total_length
+      outer_ip_header:checksum()
+   elseif backend_len == 16 then -- IPv6
+      outer_l3_type = L3_IPV6
+      outer_ip_header = IPV6:new({src= self.ipv6_addr,
+                                        dst = backend,
+                                        next_header = L4_GRE,
+                                        hop_limit = self.ttl})
+      outer_ip_header:payload_length(ip_total_length + gre_header:sizeof())
+   else
+      error("backend length must be 4 (for IPv4) or 16 (for IPv6)")
+   end
+
    datagram:push(outer_ip_header)
 
    local outer_eth_header = Ethernet:new({src = self.src_mac or eth_dst,
                                           dst = self.dst_mac,
-                                          type = L3_IPV4})
+                                          type = l3_type})
```

Comment:
This should be `outer_l3_type` not `l3_type`, right?

Reply:
On a slightly different note: we should make sure to design our tests to catch issues like this.

Reply:
Yeah, this was before I got the automated testing running. I'll fix the bug after adding a test to catch it.